File: flex+bison/tree.c
#include <stdlib.h>
#include "tree.h"

extern int yylineno;

EXP *makeEXP_identifier(char *identifier)
{
	EXP *e = malloc(sizeof(EXP));
	e->lineno = yylineno;
	e->kind = k_expressionKindIdentifier;
	e->val.identifier = identifier;
	return e;
}

EXP *makeEXP_intLiteral(int intLiteral)
{
	EXP *e = malloc(sizeof(EXP));
	e->lineno = yylineno;
	e->kind = k_expressionKindIntLiteral;
	e->val.intLiteral = intLiteral;
	return e;
}

EXP * makeEXP_binary(ExpressionKind op, EXP *lhs, EXP *rhs)
{
	EXP *e = malloc(sizeof(EXP));
	e->lineno = yylineno;
	e->kind = op;
	e->val.binary.lhs = lhs;
	e->val.binary.rhs = rhs;
	return e;
}

File: ocamllex+menhir/tree.ml
type operator =
  | Plus
  | Minus
  | Mult
  | Div
;;

type exp =
  (* id string, line number *)
  | Identifier of string * int
  (* literal value, line number *)
  | IntLiteral of int * int
  (* lhs, rhs, operator *)
  | Binary of exp * exp * operator
;;

type program =
  Program of exp
;;